### PR TITLE
BadGuy Squish Particle Effects & Enhance Particle Physics

### PR DESCRIPTION
**Pull request by lilinitsy**
Right now, this does two things
- particles can now optionally have a "gravitational" effect, or an acceleration in the y direction to make them more interesting.
- Badguy's now emit particles when they're killed by a squish, to make it a little more interesting.

TODO:
- Maybe add a more complicated subclass of Particles, PhysicsParticles, which will allow for more complicated particle movement (think effecting x, y, at t^2 and t^3 rates)
~~- Perhaps add in a "z" value to give depth to the particles and increasee / decrease size accordingly.~~ Done without storing a "z" value, just decreased the size over time.
- Particles are usually pretty gray or white... change that?
- Add different particle effects for different badguys.

### CODE REVIEW

File: src/math/vector.hpp [hunk 1]
```diff
@@ -116,6 +116,11 @@ class Vector
     return Vector(floorf(x), floorf(y));
   }
 
+  float magnitude()
```
Comment (christ2go) [inline src/math/vector.hpp]: Can you use ``norm`` instead?

Comment (lilinitsy) [reply src/math/vector.hpp]: Didn't even notice that. Sure thing.

**Comment by tobbi** [comment]
Hey @lilinitsy Could you provide a youtube video of this? I currently do not have time to compile it myself.

**Comment by lilinitsy** [comment]
@tobbi here's a low resolution gif https://giphy.com/gifs/5hiCfND58cHi4RCzt1/fullscreen.
Thanks @qwertychouskie, I'd missed that one. Handled.

**Comment by Grumbel** [comment]
[Video of it in action](https://www.youtube.com/watch?v=w_6p1s6TMFA)

Doesn't look much good in its current form and the fillrect based particles are rather ugly to begin with.

I wouldn't be opposed to some reaction when jumping on enemies, but it needs to be more subtle and preferably sprite/surface based.

**Comment by lilinitsy** [comment]
Feel free to tweak it.

**Comment by tobbi** [comment]
I'll close this mainly because of https://github.com/SuperTux/supertux/pull/904#issuecomment-434519997

**Pull request closed without merging**